### PR TITLE
Make ActionListener.wrap(Runnable) more Predictable

### PR DESCRIPTION
**Pull request by original-brownbear**
The current implementation behaved in a non-obvious manner.
If the runnable throws but is executed in `onResponse` then it will
actually be executed twice and the exception be lost.
This means that the exception during the first call will simply be quietly swallowed.
in case the runnable (as is often the case) is a `RunOnce` or similar, where the second
execution in `onFailure` is a noop this means any exception is completely swallowed.

Mainly motivated by how tricky it was to reason through possible failure modes in resource releasing around the searchable snapshots cache.

Relates efforts like #69420

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 1]
```diff
@@ -232,7 +232,33 @@ public String toString() {
      * @return a listener that listens for responses and invokes the runnable when received
      */
     static <Response> ActionListener<Response> wrap(Runnable runnable) {
-        return wrap(r -> runnable.run(), e -> runnable.run());
+        return new ActionListener<>() {
+            @Override
+            public void onResponse(Response response) {
+                try {
+                    runnable.run();
+                } catch (Exception e) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java:240]: nit: `RuntimeException`

File: server/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 2]
```diff
@@ -232,7 +232,33 @@ public String toString() {
...
+                    throw e;
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                try {
+                    runnable.run();
+                } catch (Exception ex) {
+                    ex.addSuppressed(e);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java:251]: nit: `RuntimeException`.

**Review by henningandersen — APPROVED** [review]
LGTM, nice catch!

I wonder if we can add a small test to verify the behavior?

**Comment by original-brownbear** [comment]
Thanks Henning!

> I wonder if we can add a small test to verify the behavior?

it's a little tricky right? We could only test the assertion which seems weird?

**Pull request merged**